Review

den7466 opened the pull request:
Отправка на ревью

den7466 committed:
Добавлен функционал открытия и закрытия боковой панели. Закрытие панели по нажатию на оверлей.

den7466 committed:
Добавлен функционал открытия и закрытия боковой панели. Закрытие панели по нажатию на оверлей.

den7466 committed:
Реализован весь функционал. Отправка на Первое ревью.

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,196 @@
+		backgroundColor: backgroundColors[0],
+		contentWidth: contentWidthArr[0],
+	});
+	const [dataPage, setDataPage] = useState<TDataState>({
```
**Надо исправить:**
1. Это единственный стейт, который должен остаться в компоненте `App`.
Каждый компонент сам должен отвечать за хранение и обработку своего состояния. Использовать подъем состояния в родительский компонент не имеет смысла для нашего приложения. Состояние формы нужно описать в самом компоненте `ArticleParamsForm`, а не передавать его из `App`, т.к. это состояние описывает именно `ArticleParamsForm`. В состоянии `App` должны хранится только данные о текущем отображении статьи.

2. Настройки статьи по умолчанию описаны в объекте `defaultArticleState`  в файле `constants/articleProps.ts` стартеркита. В проекте нужно использовать значения из этого объекта как настройки по умолчанию, как для самой статьи, так и как начальные значения полей форм:
```
const [ dataPage, setDataPage ] = useState(defaultArticleState);
```

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,196 @@
 				} as CSSProperties
 			}>
-			<ArticleParamsForm />
+			<ArticleParamsForm>
```
**Надо исправить:**
В результате исправления замечания про то, что стейт формы и все обработчики, связанные с формой должны "переехать" в `ArticleParamsForm` вызов этого компонента в `App` выглядит приблизительно так:
```
<ArticleParamsForm setDataPage={setDataPage} />
```
или так:
```
<ArticleParamsForm dataPage={dataPage} setDataPage={setDataPage} />
```
зависит от вашего решения

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,196 @@
+		e.preventDefault();
+		setDataForm({
+			...dataForm,
+			font: fontFamilyOptions[0],
```
**Надо исправить:**
В проекте нужно использовать значения из объекта `defaultArticleState` как настройки по умолчанию, как для самой статьи, так и как начальные значения полей форм.

Не допускается использовать для значения по умолчанию элементы с нулевым индексом из массивов свойств. То что настройки по умолчанию имеют именно нулевой индекс можно считать просто совпадением и это может измениться в любой момент.

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,196 @@
+		});
+	};
+
+	const handleOnSubmit = (e: React.BaseSyntheticEvent | SubmitEvent) => {
```
**Надо исправить:**
В React для типизации события  формы лучше использовать тип `FormEvent<HTMLFormElement>` так как он обеспечивает специфичность, безопасность и доступ к необходимым свойствам и методам для работы с формами.

olgalatkina commented on src/components/arrow-button/ArrowButton.tsx:
```diff
@@ -1,19 +1,38 @@
 			tabIndex={0}
-			className={styles.container}>
-			<img src={arrow} alt='иконка стрелочки' className={styles.arrow} />
+			className={clsx({
```
**Отлично:**
Замечательно, что для условного добавления css-классов используется библиотека `clsx`.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.module.scss:
```diff
@@ -16,6 +16,7 @@
 	display: flex;
 	flex-shrink: 0;
 	flex-direction: column;
+	gap: 50px;
```
Супер! Оптимальное решение.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,60 @@
+				})}>
+				<form
+					className={styles.form}
+					onSubmit={handleOnSubmit}
```
**Отлично:**
Отлично, что обработчик сохранения данных вызывается при отправке формы.

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,196 @@
+	contentWidth: OptionType;
+};
+
 const App = () => {
```
**Можно лучше:**
Обычно корневой `index.tsx` держат максимально компактным. Стоит вынести компонент `App` в отдельный файл в папке `components/app/App.tsx`.

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,196 @@
+
 	return (
 		<div
 			className={clsx(styles.main)}
```
**Можно лучше:**
Для одиночных селекторов не стоит использовать библиотеку `clsx`, лучше исправить этот момент здесь и в компоненте `Article`.

olgalatkina reviewed (changes requested):
Здравствуйте.

**У вас получилась хорошая работа:**
- Полностью реализован функционал.
- Для условного добавления css-классов спользуется библиотека `clsx`.
- Обработчик сохранения данных вызывается при отправке формы.

**Но есть несколько моментов, которые надо поправить:**
- [x] Состояние формы нужно описать в самом компонент `ArticleParamsForm`.
- [x] Для настроек по умолчанию лучше использовать `defaultArticleState`.
- [x] Немного поправить типизацию.

**Правки можно лучше остаются на ваше усмотрение:**
- [x] Вынести компонент `App` в отдельный файл.
- [x] Лучше не использовать библиотеку `clsx` для одиночных селекторов.

Успешного рефакторинга!

den7466 committed:
fix(articleparamsform, app, index): перенесен ArticleParamsForm, вынесен отдельно App, мелкие фиксы

Состояние формы описано в самом компоненте ArticleParamsForm, для настроек по умолчанию использован
defaultArticleState, поправлена типизация, вынесен компонент App в отдельный файл, убрано
использование clsx для одиночных селекторов.

Review #1

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,43 +1,85 @@
+		setDataPage(defaultArticleState);
+	};
+
+	useEffect(() => {
```
**Надо исправить:**
Вешать обработчик нужно при открытии сайдбара и удалять при закрытии. Для реализации этого в массиве зависимостей должен быть указан флаг управления отображением оверлея и в `useEffect` добавлена проверка, что оверлей открыт, и только в этом случае навешивать обработчики.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,43 +1,85 @@
+	const panelRef = useRef<HTMLElement | null>(null);
+	const arrowRef = useRef<HTMLDivElement | null>(null);
+
+	const handleTogglePanel = (e: React.BaseSyntheticEvent | MouseEvent) => {
```
**Надо исправить:**
1. `e.stopPropagation();` не нужен.
2. Условие можно и нужно записать проще: `setStateIsOpen(!isOpen);`
Упрощение логики за счет прямого переключения состояния на основе текущего значения снижает необходимость в условных проверках, что приводит к более чистому и эффективному коду.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,43 +1,85 @@
+	const [dataForm, setDataForm] =
+		useState<ArticleStateType>(defaultArticleState);
+	const panelRef = useRef<HTMLElement | null>(null);
+	const arrowRef = useRef<HTMLDivElement | null>(null);
```
**Надо исправить:**
Этот реф не понадобится. В связи с этим обработчик `handleClosePanelWithOverlay` нужно будет поправить.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,43 +1,85 @@
-		backgroundColorSelect: React.ReactNode;
-		contentWidthSelect: React.ReactNode;
-	};
+	setDataPage: React.Dispatch<React.SetStateAction<ArticleStateType>>;
```
Как вариант, тип для `setDataPage` можно записать так:
```
(newDataPage: ArticleStateType) => void;
```

olgalatkina reviewed (changes requested):
Отличная работа!

**Осталось поправить совсем немного:**
1. Не понадобится реф для кнопки.
2. Можно обойтись без использования `e.stopPropagation()`.
3. Переключать состояние надо на основе текущего значения.
4. Нужно добавить `useEffect` массив зависимостей и предотвратить навешивание обработчика, если сайдбар закрыт.

Успешного рефакторинга!

den7466 committed:
fix(articleparamsform, arrowbutton): исправлены замечания, отправка на Третье ревью

Удален реф для кнопки, убрано e.stopPropagation() из handleTogglePanel, переделано переключение
сайдбара на основе текущего состояния, переделан useEffect - добавлена проверка что оверлей открыт,
изменена типизация setDataPage.

olgalatkina reviewed (approved):
Добрый день!

**Отличная работа! проект принимается**🎉

Успехов в дальнейшем обучении!
